# ==============================================================================
# REGRESSOR COMMIT DIFF
# ==============================================================================
# File: toolkit/content/widgets/browser-custom-element.js
# Commit: 4b76024db21c
# Full Hash: 4b76024db21cff006a7424a35c412c6e0a153eb0
# Author: Kris Maglione <maglione.k@gmail.com>
# Date: 2020-09-22 03:47:42
# Regressor Bug: 1655866
# File Overlap Count: 1
# Description:
#   Bug 1655866: Part 5 - Use native PermitUnload implementation from front-end code. r=nika,mconley
#   
#   Differential Revision: https://phabricator.services.mozilla.com/D88318
# ==============================================================================

diff -r 55fb069319f1 -r 4b76024db21c toolkit/content/widgets/browser-custom-element.js
--- a/toolkit/content/widgets/browser-custom-element.js	Mon Sep 21 22:41:01 2020 +0000
+++ b/toolkit/content/widgets/browser-custom-element.js	Mon Sep 21 22:41:04 2020 +0000
@@ -26,12 +26,6 @@
 
   ChromeUtils.defineModuleGetter(
     LazyModules,
-    "PermitUnloader",
-    "resource://gre/actors/BrowserElementParent.jsm"
-  );
-
-  ChromeUtils.defineModuleGetter(
-    LazyModules,
     "E10SUtils",
     "resource://gre/modules/E10SUtils.jsm"
   );
@@ -55,6 +49,11 @@
     "fission.sessionHistoryInParent",
     false
   );
+  XPCOMUtils.defineLazyPreferenceGetter(
+    lazyPrefs,
+    "unloadTimeoutMs",
+    "dom.beforeunload_timeout_ms"
+  );
 
   const elementsToDestroyOnUnload = new Set();
 
@@ -85,6 +84,8 @@
       this._characterSet = null;
       this._documentContentType = null;
 
+      this._inPermitUnload = new WeakSet();
+
       /**
        * These are managed by the tabbrowser:
        */
@@ -1701,7 +1702,10 @@
           aCallback(false);
           return;
         }
-        aCallback(LazyModules.PermitUnloader.inPermitUnload(this.frameLoader));
+
+        aCallback(
+          this._inPermitUnload.has(this.browsingContext.currentWindowGlobal)
+        );
         return;
       }
 
@@ -1712,25 +1716,75 @@
       aCallback(this.docShell.contentViewer.inPermitUnload);
     }
 
-    permitUnload(aPermitUnloadFlags) {
+    async asyncPermitUnload(action) {
+      let wgp = this.browsingContext.currentWindowGlobal;
+      if (this._inPermitUnload.has(wgp)) {
+        throw new Error("permitUnload is already running for this tab.");
+      }
+
+      this._inPermitUnload.add(wgp);
+      let timeout;
+      try {
+        let result = await Promise.race([
+          new Promise(resolve => {
+            timeout = setTimeout(() => {
+              resolve({ permitUnload: true, timedOut: true });
+            }, lazyPrefs.unloadTimeoutMs);
+          }),
+          wgp
+            .permitUnload(action)
+            .then(permitUnload => ({ permitUnload, timedOut: false })),
+        ]);
+
+        return result;
+      } finally {
+        this._inPermitUnload.delete(wgp);
+        clearTimeout(timeout);
+      }
+    }
+
+    get hasBeforeUnload() {
+      function hasBeforeUnload(bc) {
+        if (bc.currentWindowContext?.hasBeforeUnload) {
+          return true;
+        }
+        return bc.children.some(hasBeforeUnload);
+      }
+      return hasBeforeUnload(this.browsingContext);
+    }
+
+    permitUnload(action) {
       if (this.isRemoteBrowser) {
-        if (!LazyModules.PermitUnloader.hasBeforeUnload(this.frameLoader)) {
+        if (!this.hasBeforeUnload) {
           return { permitUnload: true, timedOut: false };
         }
 
-        return LazyModules.PermitUnloader.permitUnload(
-          this.frameLoader,
-          aPermitUnloadFlags
+        let result;
+        let success;
+
+        this.asyncPermitUnload(action).then(
+          val => {
+            result = val;
+            success = true;
+          },
+          err => {
+            result = err;
+            success = false;
+          }
         );
+
+        Services.tm.spinEventLoopUntilOrShutdown(() => success !== undefined);
+        if (success) {
+          return result;
+        }
+        throw result;
       }
 
       if (!this.docShell || !this.docShell.contentViewer) {
         return { permitUnload: true, timedOut: false };
       }
       return {
-        permitUnload: this.docShell.contentViewer.permitUnload(
-          aPermitUnloadFlags
-        ),
+        permitUnload: this.docShell.contentViewer.permitUnload(),
         timedOut: false,
       };
     }
